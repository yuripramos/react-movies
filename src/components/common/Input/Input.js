import React, { Component } from "react";
import { string, func, bool, array, number, oneOfType } from "prop-types";

import { InputWrapper, InputField, Label, Prefix } from "./styles";

class Input extends Component {
  constructor(props) {
    super(props);
    this.state = {
      isFocused: false,
      emptyState: false,
      isValid: typeof props.valid === "boolean" ? props.valid : null,
      isDisabled: false
    };
  }

  componentDidUpdate(prevProps) {
    if (this.props.value === "" && this.state.isValid !== null) {
      this.setState({
        isValid: null
      });
    }

    if (prevProps.value === "" && this.props.value.length > 4) {
      this.setState({
        isValid:
          this.props.value === ""
            ? null
            : typeof this.props.valid === "function"
            ? this.props.valid()
            : this.state.isValid
      });
    }
  }

  render() {
    const {
      label,
      value,
      placeholder,
      onChange,
      onBlur,
      type,
      name,
      valid,
      tinyLabels,
      width,
      onFocus,
      prefix,
      ...props
    } = this.props;
    const { isFocused, emptyState, isValid } = this.state;
    return (
      <InputWrapper valid={emptyState ? false : isValid} width={width}>
        {prefix && (value || isFocused) && <Prefix>{prefix}</Prefix>}
        <InputField
          {...props}
          required
          onChange={e => {
            onChange && onChange(e);
            this.setState({ emptyState: e.target.value === "" });
          }}
          onFocus={e => {
            onFocus && onFocus(e);
            prefix && this.setState({ isFocused: true });
          }}
          onBlur={e => {
            onBlur && onBlur(e);
            this.setState({
              isFocused: false,
              isValid: valid ? valid() : this.state.isValid
            });
          }}
          prefix={prefix}
          type={type}
          value={value}
          name={name}
          onKeyDown={this.onKeyDown}
          placeholder={placeholder}
          isEmpty={!value}
          tinyLabels={tinyLabels}
        />
        {label && <Label tinyLabels={tinyLabels}>{label}</Label>}
      </InputWrapper>
    );
  }
}

Input.displayName = "Input";

Input.defaultProps = {
  placeholder: "",
  type: "text",
  name: "",
  label: null,
  valid: () => {},
  onFocus: null,
  onBlur: () => {},
  value: "",
  disableKeys: []
};

Input.propTypes = {
  placeholder: string,
  type: string,
  name: string,
  label: string,
  icon: string,
  tinyLabels: bool,
  value: oneOfType([number, string]),
  valid: oneOfType([bool, func]),
  onBlur: func,
  onFocus: func,
  disableKeys: oneOfType([number, string, array]),
  onChange: func,
  width: string,
  prefix: string
};

export default Input;
