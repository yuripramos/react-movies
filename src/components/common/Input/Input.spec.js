import React from "react";
import { shallow } from "enzyme";

import Input from "./Input";

describe("Input component", () => {
  it("should match snapshot", () => {
    const props = {
      onChange: () => {},
      type: "submit",
      name: "input",
      placeholder: "My input"
    };
    expect(shallow(<Input {...props} />)).toMatchSnapshot();
  });

  it("should match snapshot with label", () => {
    const props = {
      onChange: () => {},
      type: "submit",
      name: "input",
      label: "label"
    };
    expect(shallow(<Input {...props} />)).toMatchSnapshot();
  });

  it("should match snapshot with success validation", () => {
    const props = {
      onChange: () => {},
      type: "submit",
      name: "input",
      placeholder: "My input",
      valid: true
    };
    expect(shallow(<Input {...props} />)).toMatchSnapshot();
  });

  it("should match snapshot with icon", () => {
    const props = {
      onChange: () => {},
      type: "submit",
      name: "input",
      placeholder: "My input",
      valid: true,
      icon: "Icon"
    };
    expect(shallow(<Input {...props} />)).toMatchSnapshot();
  });

  it("should match snapshot with error validation", () => {
    const props = {
      onChange: () => {},
      type: "submit",
      name: "input",
      placeholder: "My input",
      valid: false
    };
    expect(shallow(<Input {...props} />)).toMatchSnapshot();
  });

  it("should match snapshot with no validation", () => {
    const props = {
      onChange: () => {},
      type: "submit",
      name: "input",
      placeholder: "My input",
      valid: null
    };
    expect(shallow(<Input {...props} />)).toMatchSnapshot();
  });
});
