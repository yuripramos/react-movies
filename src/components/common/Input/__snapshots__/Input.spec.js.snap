// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input component should match snapshot 1`] = `
<styled.div
  valid={null}
>
  <styled.input
    disableKeys={Array []}
    isEmpty={true}
    name="input"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    placeholder="My input"
    required={true}
    type="submit"
    value=""
  />
</styled.div>
`;

exports[`Input component should match snapshot with error validation 1`] = `
<styled.div
  valid={false}
>
  <styled.input
    disableKeys={Array []}
    isEmpty={true}
    name="input"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    placeholder="My input"
    required={true}
    type="submit"
    value=""
  />
</styled.div>
`;

exports[`Input component should match snapshot with icon 1`] = `
<styled.div
  valid={true}
>
  <styled.input
    disableKeys={Array []}
    icon="Icon"
    isEmpty={true}
    name="input"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    placeholder="My input"
    required={true}
    type="submit"
    value=""
  />
</styled.div>
`;

exports[`Input component should match snapshot with label 1`] = `
<styled.div
  valid={null}
>
  <styled.input
    disableKeys={Array []}
    isEmpty={true}
    name="input"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    placeholder=""
    required={true}
    type="submit"
    value=""
  />
  <styled.label>
    label
  </styled.label>
</styled.div>
`;

exports[`Input component should match snapshot with no validation 1`] = `
<styled.div
  valid={null}
>
  <styled.input
    disableKeys={Array []}
    isEmpty={true}
    name="input"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    placeholder="My input"
    required={true}
    type="submit"
    value=""
  />
</styled.div>
`;

exports[`Input component should match snapshot with success validation 1`] = `
<styled.div
  valid={true}
>
  <styled.input
    disableKeys={Array []}
    isEmpty={true}
    name="input"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    placeholder="My input"
    required={true}
    type="submit"
    value=""
  />
</styled.div>
`;
